### ISSUE TITLE
[feature] Handling fine control of builds: C[XX]FLAGS, etc. 

### ISSUE THREAD

**Issue by ruilvo** [conan-io/conan#10805]
Following a [discussion on the cpplang slack](https://cpplang.slack.com/archives/C41CWV9HA/p1646556536574179), I'm moving discussion here.

The discussion more or less goes like this:

I'm the author of library L1.
In my library L1, I want to use dependencies D1, D2, D3.
D1, D2, D3 consumed via Conan, of course using a `conanfile.[txt|py]`, whatever would be more convenient. 
CMake for building L1.
For whatever reason, it would be beneficial to build D# with flag `-flagD#`.

How can this be achieved with Conan?

Conan doesn't have a mechanism to pass arbitrary additional compilation flags, or to set environment variables for dependencies.

And this is where the problem starts.

Assuming the dependency's build system respects the `CXXFLAGS` environment variable, we could go with that route. But:

- Conan doesn't take `CXXFLAGS` into account when building package IDs, so without `--build always` it would just pick binaries from the repository anyway;
- There isn't a straightforward way to set this environment variable anyway, leading to the following discussion:

The first suggestion was to use profiles, shipped alongside the library L1, but that's a big can of worms:

Profiles would require me to set this section:
```
[settings]
os=
arch=
compiler=
compiler.version=
```
for every compiler I expect users/devs of L1 to use. That's pretty much impossible.
Another option is to _not_ do that and have a simpler profile with just the `[env]` section filled in, but since profiles can't be "merged" at "usage time", users/devs of L1 can't use their compiler settings when `conan [create|install]`-ing:
```
conan install -s compiler="Visual Studio" -s compiler.version="16"  # etc etc etc, even with a profile...
```

Compare this with, for example, how CMake does it:
```cmake
if(MSVC)
  add_compile_options("/fp:fast")
else()
  add_compile_options("-ffast-math")
endif()
```
The `else()` clause is pretty much guaranteed to work on pretty much "all" GCC or Clang versions, easily. 
This also applies to dependencies: For example, if I used `git submodule` + `add_subdirectory`, or even an `ExternalContent` /`FetchContent` workflow  it would be possible to control the way `CMAKE_CXX_FLAGS`  would be initialized indeed for those dependencies.

I guess an option would be to have a `extra_cxx_flags` option for all recipes. If you ever want to consider doing this to CCI, count me in to help. For example, an `extra_cxx_flags` option could translate in `f"-DCMAKE_CXX_FLAGS={extra_cxx_flags}"` argument passed via the CMake build helper. 

The _really_ high level option would be to have something like a `mydependency1.cxxflags=["-ffast-math"] #pseudocode` setting in a consumer `conanfile.txt`, but that's a high dream. 
Imagine this:
```
self.dependencies.["D1"].cxx_flags = ["-f1"] # iterable for append, literal for replace, like profiles do
```

Read the slack discussion for more context, or feel free to ask questions. 

In conclusion, I'd like a system to be able to have this fine control over how my dependencies are built.

**Comment by ruilvo** [comment]
An interesting discussion I had about this issue:

```
 10:28 <LordKalma> honestly, the "right" answer for that would honestly to have a directory in the project with custom conan recipes
10:29 <LordKalma> and orchestrate "conan create" with a super-build
10:32 <very_sneaky> ehhhhh.. idk about that
10:32 <very_sneaky> that breaks encapsulation i think
10:33 <LordKalma> welcome to the hell dependencies are :)
10:33 <very_sneaky> maybe in the `requires()` call you can set env vars for the upstream build
10:33 <LordKalma> you can't because they aren't considered in the package ID and the only way to pass them in is with profiles anyway
10:33 <LordKalma> Conan just doesn't support this workflow
10:33 <very_sneaky> i.e. `requires("openssl/1.1.1l", env={"CXXFLAGS": ["-ffast-math"]})
10:34 <LordKalma> if you want to use a dependency out-of-the-box, it's fine. Anything clever is out of the table
10:34 <LordKalma> very_sneaky, that would require having the env part of the package ID, which 1) isn't implemented, 2) it's not easy to implement, 3) would be shit
10:34 <LordKalma> 3) because you don't want for example PATH to change your package ID
10:35 <LordKalma> only like C[XX]FLAGS and the linker equivalents
10:35 <very_sneaky> i think the 90% solution would just be what i have above with package ID on the unique key/pair values
10:35 <LordKalma> probably
10:35 <very_sneaky> it wouldn't be perfect, because you'd get package id changing on anything
10:35 <LordKalma> ooor have all CCI recipes have an "extra_cxx_flags" option haha
10:36 <very_sneaky> but maybe better than nothing, and obviously opt in - package id changing is just a convenience thing
10:36 <very_sneaky> yeah, that would be an option too
10:36 <LordKalma> that's what I'm saying. you need to be sane and stable. You don't want to have to rebuild on every single thing
10:36 <very_sneaky> defaulted field
10:36 <very_sneaky> LordKalma: you wouldn't _have_ to
10:36 <very_sneaky> it would be opt in
10:37 <LordKalma> oh the other hand, it can also be dangerous
10:37 <LordKalma> I'm not sure what CMake does when you pass both -DCMAKE_CXX_FLAGS and have CXXFLAGS defined
10:37 <LordKalma> because with a "extra_cxx_flags" option you'd do the former and would that invalidate the latter, thus invalidating your profile?
10:38 <very_sneaky> still, a net win - you wouldn't be introducing any additional complexity
10:38 <LordKalma> "For any configuration run ( including the first), the environment variable will be ignored if the CMAKE_CXX_FLAGS variable is defined." https://cmake.org/cmake/help/latest/envvar/CXXFLAGS.html
10:38 <LordKalma> as expected
10:38 <LordKalma> very_sneaky, but it doesn't comply very well with the principle of least surprise, I'd say
10:39 <very_sneaky> maybe, you could do some validation
10:39 <LordKalma> https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS_INIT.html#variable:CMAKE_%3CLANG%3E_FLAGS_INIT
10:39 <very_sneaky> i'm not proposing this as the best sollution
10:39 <LordKalma> this could be cool tough
10:39 <very_sneaky> just one that would get us most of the way there
10:39 <LordKalma> CMAKE_<LANG>_FLAGS_INIT joins itself with the env variable
10:39 <LordKalma> so that's a solution for cmake-based recipes, hum...
```

**Comment by memsharded** [comment]
We are investigating the possibility of a explicit management of compile flags, something that was not done before (it was kind of possible via environment, but not managed directly by Conan).

https://github.com/conan-io/conan/pull/10800 is adding:

- The possibility of defining ``[conf]`` values in your profile and command line for CFLAGS, CXXFLAGS
- The builtin toolchains (like ``CMakeToolchain``, etc) will read this conf and map those flags to the build system scripts (like ``conan_toolchain.cmake``, in the CMAKE_<LANG>_FLAGS_INIT)
- The ``[conf]`` has possibility to target per-package if desired like ``<pkg-pattern>:tools.build:cxxflags=value``
- ``conf_info`` can also be defined in tool_requires packages that encapsulate toolchains to be automatically applied to consumers of those tool_requires

**Comment by SSE4** [comment]
one missing peace, which might be (extremely) hard to nail.
these flags do not participate in package ID.
it means, whenever you specify `tools.build:cxxflags=-fsanitize-address`, conan will not rebuild binaries by default, but will just download binaries from the remote, which more likely were not compiled with `-fsanitize-address`. not something one expect.
so, user will have to enfore `--build always` (or at least something to avoid getting binaries from remote). other alternative could be a hook checking `conaninfo.txt` of packages downloaded.

next, imagine we had `cxxflags` to be encoded into the package ID.
we already have toolchain packages that set these flags (e.g. Android NDK).
therefore, as soon as conan sees a package modifying flags, it should rebuild a graph, because package ID changes.
then we will have a vicious circle, because of conditionals in recipes, like:
```
if "-fsanitize-address" in self.cxxflags:
    self.requires.add("sanitizer-libs/1.0")
```
but new library may again modify the flags:
```
if "sanitizer-libs" in self.dependencies:
    self.cpp_info.cxxflags.append("-fsanitize-address")
```
which could lead to a vicious circle in graph build process. you need to know flags to build the graph, but you need to build the graph to know the flags.

**Comment by memsharded** [comment]
> one missing peace, which might be (extremely) hard to nail.
these flags do not participate in package ID.

There are plans to make ``conf`` optionally part of the package_id. There is already some undocumented infrastructure.
But lets go one step at a time. The first important thing is to validate that it is practical to define and pass the flags.

Regarding the other issues, it is impossible to have a generic, automatic model for package_id regarding flags. It will be the responsibility of users, and in the general case, the recommended approach is to build high-level abstractions, for example to add model to the settings via ``settings.yml`` to represent the sanitizers, and then have the toolchains map that to actual flags.

**Comment by ruilvo** [comment]
I'm happy the team is thinking about these issues! And mostly reaching the same conclusions! 

Thanks for everything!
